Documentation/Contributors/CodingGuide/README.md

```diff
@@ -0,0 +1,662 @@
+    this._url = options.url;
+};
+
+defineProperties(Cesium3DTileset.prototype, {
+    url : {
+        get : function() {
+            return this._url;
+        }
+    }
+});
+```
+Getters can perform any needed computation to return the property, but the performance expectation is that they execute quickly.
+
+Setters can also perform computation before assigning to a private property, set a flag to delay computation, or both, for example:
+```javascript
+defineProperties(UniformState.prototype, {
+    viewport : {
+        get : function() {
+            return this._viewport;
+        },
+        set : function(viewport) {
+            if (!BoundingRectangle.equals(viewport, this._viewport)) {
+                BoundingRectangle.clone(viewport, this._viewport);
+
+                var v = this._viewport;
+                var vc = this._viewportCartesian4;
+                vc.x = v.x;
+                vc.y = v.y;
+                vc.z = v.width;
+                vc.w = v.height;
+
+                this._viewportDirty = true;
+            }
+        }
+    }
+});
+```
+
+* :speedboat: Calling the getter/setter function is slower than direct property access so functions internal to a class can use the private property directly when appropriate.
+
+### Shadowed Property
+
+When the overhead of getter/setter functions is prohibitive or reference type semantics are desired, e.g., the ability to pass a property as a `result` parameter so its properties can be modified, consider combining a public property with a private shadowed property, e.g.,
+```javascript
+var Model = function(options) {
+    this.modelMatrix = Matrix4.clone(defaultValue(options.modelMatrix, Matrix4.IDENTITY));
+    this._modelMatrix = Matrix4.clone(this.modelMatrix);
+};
+
+Model.prototype.update = function(frameState) {
+    if (!Matrix4.equals(this._modelMatrix, this.modelMatrix)) {
+        // clone() is a deep copy. Not this._modelMatrix = this._modelMatrix
+        Matrix4.clone(this.modelMatrix, this._modelMatrix);
+
+        // Do slow operations that need to happen when the model matrix changes
+    }
+};
+```
+
+### Put the Constructor Function at the Top of the File
+
+It is convenient for the constructor function to be at the top of the file even if it requires that helper functions rely on **hoisting**, for example, `Cesium3DTileset.js`,
+```javascript
+function loadTilesJson(tileset, tilesJson, done) {
+    // ...
+}
+
+var Cesium3DTileset = function(options) {
+    // ...
+    loadTilesJson(this, options.url, function(data) {
+       // ...
+    });
+};
+```
+is better written as
+```javascript
+var Cesium3DTileset = function(options) {
+    // ...
+    loadTilesJson(this, options.url, function(data) {
+       // ...
+    });
+};
+
+function loadTilesJson(tileset, tilesJson, done) {
+    // ...
+}
+```
+even though it relies on implicitly hoisting `loadTilesJson` to the top of the file.
+
+Cesium never relies on hoisting variables.
```

Comment:
You already mentioned this in another section, I would just take this line out or change it to "While we avoid variable hoisting, function hoisting like this is allowed"